src/PooledArrays.jl

```diff
@@ -165,31 +186,120 @@ Base.size(pa::PooledArray) = size(pa.refs)
 Base.length(pa::PooledArray) = length(pa.refs)
 Base.lastindex(pa::PooledArray) = lastindex(pa.refs)
 
-Base.copy(pa::PooledArray) = PooledArray(RefArray(copy(pa.refs)), copy(pa.invpool))
-# TODO: Implement copy_to()
+Base.copy(pa::PooledArray) = PooledArray(pa)
+
+Base.copyto!(dest::PooledArray{T, R, N, RA},
+             src::PooledArray{T, R, N, RA}) where {T, R, N, RA} =
+    copyto!(dest, 1, src, 1, length(src))
+
+# TODO: handle case when dest is SubArray
+
+Base.copyto!(dest::PooledArray{T, R, N, RA},
+             src::SubArray{T, N, PooledArray{T, R, M, RA}}) where {T, R, N, M, RA} =
+    copyto!(dest, 1, src, 1, length(src))
+
+function Base.copy!(dest::PooledArray{T, R, N, RA},
+                    src::PooledArray{T, R, N, RA}) where {T, R, N, RA}
+    copy!(dest.refs, src.refs)
+    Threads.atomic_add!(src.refcount, 1)
+    dest.pool = src.pool
+    dest.invpool = src.invpool
+    dest.refcount = src.refcount
+    return dest
+end
+
+function Base.copy!(dest::PooledArray{T, R, N, RA},
+                    src::SubArray{T, N, PooledArray{T, R, M, RA}}) where {T, R, N, M, RA}
+    copy!(dest.refs, view(parent(src).refs, src.indices))
+    Threads.atomic_add!(src.refcount, 1)
+    dest.pool = src.pool
+    dest.invpool = src.invpool
+    dest.refcount = src.refcount
+    return dest
+end
+
+function Base.copyto!(dest::PooledArray{T, R, N, RA}, doffs::Union{Signed, Unsigned},
+                 src::PooledArray{T, R, N, RA}, soffs::Union{Signed, Unsigned},
+                 n::Union{Signed, Unsigned}) where {T, R, N, RA}
+    n == 0 && return dest
+    n > 0 || Base._throw_argerror()
+    if soffs < 1 || doffs < 1 || soffs+n-1 > length(src) || doffs+n-1 > length(dest)
+        throw(BoundsError())
+    end
+
+    if length(dest.pool) == 0
```

Comment:
Maybe a code comment explaining this if branch of code?

Reply:
I have added a comment - it is useful if we want to use a pattern like:
```
newpa = similar(pa, newsize)
copyto!(newpa, start1, pa, start2, n)
```

The only case we do not resolve is line like
https://github.com/JuliaData/DataFrames.jl/pull/2622/files#diff-00fe5953fad7e5572a8a4ae57ab3724fd3a3895cae501c40bc60a9fed12cf60fR221

as here we have a problem that union with `Missing` is potentially introduced and we have to copy a pool anyway (because we add `Missing` to it). This kind of problem could probably be resolved by adding special handling of `Missing` in PolledArrays.jl like it is done in CategoricalArrays.jl, but I think it should be left for a separate PR.

The conclusion is that `innerjoin` will be always fully fast after this PR, but `leftjoin`, `rightjoin` and `outerjoin` will not be maximally fast because of this union-taking with `Missing`.